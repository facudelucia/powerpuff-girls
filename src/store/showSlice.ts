import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";
import { Show, Episode } from "../types/showTypes";

export interface ShowState {
  show: Show | null;
  episodes: Episode[];
  selectedEpisode: Episode | null;
  status: "idle" | "loading" | "failed";
}

const initialState: ShowState = {
  show: null,
  episodes: [],
  selectedEpisode: null,
  status: "idle",
};

export const fetchShowDetails = createAsyncThunk(
  "show/fetchShowDetails",
  async () => {
    const response = await axios.get("https://api.tvmaze.com/shows/6771");
    return response.data;
  }
);

export const fetchEpisodes = createAsyncThunk(
  "show/fetchEpisodes",
  async () => {
    const response = await axios.get(
      "https://api.tvmaze.com/shows/6771/episodes"
    );
    return response.data;
  }
);

const showSlice = createSlice({
  name: "show",
  initialState,
  reducers: {
    selectEpisode: (state, action) => {
      state.selectedEpisode =
        state.episodes.find(
          (episode: Episode) => episode.id === action.payload
        ) || null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchShowDetails.pending, (state) => {
        state.status = "loading";
      })
      .addCase(fetchShowDetails.fulfilled, (state, action) => {
        state.status = "idle";
        state.show = action.payload;
      })
      .addCase(fetchEpisodes.fulfilled, (state, action) => {
        state.episodes = action.payload;
      });
  },
});

export const { selectEpisode } = showSlice.actions;
export default showSlice.reducer;
