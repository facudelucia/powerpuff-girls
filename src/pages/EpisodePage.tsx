import React, { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { useParams, useNavigate  } from 'react-router-dom';
import { selectEpisode, fetchEpisodes } from '../store/showSlice';
import EpisodeDetails from '../components/EpisodeDetails';
import { AppDispatch } from '../store';
import styled from 'styled-components';

const PageContainer = styled.div`
  padding: 40px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  min-height: 100vh;
  background: linear-gradient(135deg, #ece9e6, #ffffff);

  @media (max-width: 768px) {
    padding: 20px;
  }
`;

const EpisodeCard = styled.div`
  background-color: white;
  padding: 30px;
  border-radius: 10px;
  box-shadow: 0 4px 20px rgba(0, 0, 0, 0.1);
  max-width: 1000px;
  width: 100%;
  transition: transform 0.3s ease, box-shadow 0.3s ease;

  &:hover {
    box-shadow: 0 6px 25px rgba(0, 0, 0, 0.15);
  }

  @media (max-width: 768px) {
    padding: 20px;
  }
`;

const GoBackButton = styled.button`
  background-color: #3498db;
  color: white;
  padding: 10px 20px;
  font-size: 1rem;
  font-weight: bold;
  border: none;
  border-radius: 5px;
  cursor: pointer;
  margin-bottom: 20px;
  transition: background-color 0.3s ease, transform 0.2s ease;

  &:hover {
    background-color: #2980b9;
    transform: scale(1.05);
  }

  &:focus {
    outline: none;
  }
`;

const EpisodePage: React.FC = () => {
  const { id } = useParams<{ id: string }>();
  const dispatch: AppDispatch = useDispatch();
  const navigate = useNavigate(); 

  useEffect(() => {
      dispatch(selectEpisode(Number(id)));
  }, [dispatch, id]); 
  
  const handleGoBack = () => {
    navigate(-1);
  };

  return (
    <PageContainer>
      <EpisodeCard data-testid="episode-details">
        <GoBackButton onClick={handleGoBack}>{"< Go Back"}</GoBackButton>
        <EpisodeDetails />
      </EpisodeCard>
    </PageContainer>
  );
};

export default EpisodePage;
