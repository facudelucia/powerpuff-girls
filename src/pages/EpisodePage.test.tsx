import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import EpisodePage from './EpisodePage';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { store } from '../store';

describe('EpisodePage Component', () => {
  test('renders EpisodePage and shows loading state', async() => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <EpisodePage />
        </MemoryRouter>
      </Provider>
    );

    await waitFor(() => {
      expect(screen.getByTestId('loading')).toBeInTheDocument();
    });
  });

  test('matches the snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <EpisodePage />
        </MemoryRouter>
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
