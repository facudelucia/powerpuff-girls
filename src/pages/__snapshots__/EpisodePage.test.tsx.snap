// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodePage Component matches the snapshot 1`] = `
<div
  className="sc-iKTcqh gXHfRt"
>
  <div
    className="sc-gnpbhQ krRGqB"
    data-testid="episode-details"
  >
    <button
      className="sc-la-DxNn bcarGq"
      onClick={[Function]}
    >
      &lt; Go Back
    </button>
    <div
      className="sc-dsLQwm hzXjGX"
      data-testid="loading"
    />
  </div>
</div>
`;
